qiskit_ibm_runtime/utils/json.py

```diff
@@ -264,24 +273,47 @@ def default(self, obj: Any) -> Any:  # pylint: disable=arguments-differ
                 ),  # type: ignore[no-untyped-call]
             )
             return {"__type__": "Instruction", "__value__": value}
-        # TODO proper way to do this?
-        if isinstance(obj, EstimatorPub):
-            return {
-                "circuit": obj.circuit,
-                "observables": obj.observables,
-                "parameter_values": obj.parameter_values,
-                "precision": obj.precision,
-            }
         if isinstance(obj, ObservablesArray):
-            return obj.tolist()
+            return {"__type__": "ObservablesArray", "__value__": obj.tolist()}
         if isinstance(obj, BindingsArray):
             out_val = {"shape": obj.shape}
             encoded_data = {}
             for key, val in obj.data.items():
                 encoded_data[json.dumps(key, cls=RuntimeEncoder)] = val
             out_val["data"] = encoded_data
             return {"__type__": "BindingsArray", "__value__": out_val}
-
+        if isinstance(obj, BitArray):
+            out_val = {"array": obj.array, "num_bits": obj.num_bits}
+            return {"__type__": "BitArray", "__value__": out_val}
+        if isinstance(obj, DataBin):
+            out_val = {
+                "field_names": obj._FIELDS,
+                "field_types": [str(field_type) for field_type in obj._FIELD_TYPES],
+                "shape": obj._SHAPE,
+                "values": {field_name: getattr(obj, field_name) for field_name in obj._FIELDS},
+            }
+            return {"__type__": "DataBin", "__value__": out_val}
+        if isinstance(obj, EstimatorPub):
+            out_val = {
+                "circuit": obj.circuit,
+                "observables": obj.observables,
+                "parameter_values": obj.parameter_values,
+                "precision": obj.precision,
+            }
+            return {"__type__": "EstimatorPub", "__value__": out_val}
+        if isinstance(obj, SamplerPub):
+            out_val = {
+                "circuit": obj.circuit,
+                "parameter_values": obj.parameter_values,
+                "shots": obj.shots,
+            }
+            return {"__type__": "SamplerPub", "__value__": out_val}
+        if isinstance(obj, PubResult):
+            out_val = {"data": obj.data, "metadata": obj.metadata}
+            return {"__type__": "PubResult", "__value__": out_val}
```

Comment:
@mberna As discussed today, we need to either make sure this is not used to encode `PubResult` from the server, or change it to not pass `__type__` etc. Since the return data from the server should just be a "normal" dictionary.